### DISCUSSION TITLE
Using DryIoc ASP.NET Core MVC (.NET 5)

### DISCUSSION THREAD

**Discussion by timrhaynes** [dadhi/DryIoc#398] [Q&A] [answered]
I am attempting to upgrade a legacy .NET Framework 4.7.2 solution, to be ASP.NET Core 5

The legacy solution used DryIoc 2.12.10. It was wrapped it in a class (called DIContainer) that was passed into the Register method of classes that were referenced in appsettings.json's System/Dependency/Containers array.

In my ASP.NET Core MVC (.NET 5) solution, I upgraded to DryIoc 4.7.5 and am hoping to retain the DIContainer wrapper approach in order to avoid rewriting a bunch of application code.

I have studied the provided DryIoc samples, which cover ASP.NET Core 3.1 and WebApi usage. Neither are 100% exactly what you would do for ASP.NET Core 5, but pretty close. I have been able to get things almost working.

One thing that frustrates me is that in my IServiceProviderFactory.CreateBuilder implementation, I cannot return DIContainer, for the Main method to use when invoking HostBuilder.Build(). I guess it makes sense, since DIContainer does not conform to the interface it's probably trying to use internally.

My current issue, though, is that in my Program.cs Main() method, I get the error: System.InvalidOperationException: 'No service for type 'Microsoft.Extensions.Hosting.IHost' has been registered.'

I'm a bit lost on how to proceed with that. I'm not sure what code people would need to see to help debug things. Please feel free to ask for specifics, and I'll try to explain further, with code samples.

Thank you for any help!

**Comment by dadhi** [comment, marked as answer]
> One thing that frustrates me is that in my IServiceProviderFactory.CreateBuilder implementation, I cannot return DIContainer, for the Main method to use when invoking HostBuilder.Build()

I am not sure about this problem, because HostBuilder still has the UseServiceFactoryProvider method:
https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.hosting.hostbuilder.useserviceproviderfactory?view=dotnet-plat-ext-5.0#Microsoft_Extensions_Hosting_HostBuilder_UseServiceProviderFactory__1_Microsoft_Extensions_DependencyInjection_IServiceProviderFactory___0__
so you may return `new DryIocServiceProviderFactory(...)`.

**Reply by dadhi** [reply]
> hoping to retain the DIContainer wrapper approach in order to avoid rewriting a bunch of application code.

What does this wrapper do? Could you show it here or some fragment of it?

**Reply by timrhaynes** [reply]
Well, I've made some progress, so that may not be needed. What I'm bumping into now is that I do something like this:

```
private void RegisterThings(IDIContainer container)
 {
       container.Register<IConfiguration, Configuration>();  
       var test = container.Resolve<IConfiguration>();
 }

```
 And it works... however, in another class when I try to resolve IConfigure, I get this error:
 DryIoc.ContainerException
  HResult=0x80131509
  Message=code: Error.UnableToResolveUnknownService;
message: Unable to resolve resolution root MyNamespace.IConfiguration
  from container without scope
 with Rules with {TrackingDisposableTransients} and without {ThrowOnRegisteringDisposableTransient, VariantGenericTypesInResolvedCollection}
 with FactorySelector=SelectLastRegisteredFactory
 with Made={FactoryMethod=ConstructorWithResolvableArguments, PropertiesAndFields=<custom>}
Where no service registrations found
  and no dynamic registrations found in 0 of Rules.DynamicServiceProviders
  and nothing found in 0 of Rules.UnknownServiceResolvers
  Source=DryIoc
  StackTrace:
   at DryIoc.Throw.It(Int32 error, Object arg0, Object arg1, Object arg2, Object arg3) in /_/src/DryIoc/Container.cs:line 13621
   at DryIoc.Container.TryThrowUnableToResolve(Request request) in /_/src/DryIoc/Container.cs:line 1091
   at DryIoc.Container.DryIoc.IContainer.ResolveFactory(Request request) in /_/src/DryIoc/Container.cs:line 1070
   at DryIoc.Container.ResolveAndCache(Int32 serviceTypeHash, Type serviceType, IfUnresolved ifUnresolved) in /_/src/DryIoc/Container.cs:line 367
   at DryIoc.Container.DryIoc.IResolver.Resolve(Type serviceType, IfUnresolved ifUnresolved) in /_/src/DryIoc/Container.cs:line 356
   at DryIoc.Resolver.Resolve[TService](IResolver resolver, IfUnresolved ifUnresolved) in /_/src/DryIoc/Container.cs:line 8017

Do you have any advice on that error? Thanks!

**Reply by dadhi** [reply]
@timrhaynes That means in another class you have container without `IConfiguration` registration. Likely you did something with the container rules producing the new container - then registering configuration into the new container.

**Reply by timrhaynes** [reply]
@dadhi So you are saying that I somehow have a second container that lacks an IConfiguration registration, and it's this second container that is throwing this error?

**Reply by dadhi** [reply]
@timrhaynes Yes, my guess.
